src/ext4.c

```diff
@@ -2016,23 +2016,25 @@ int ext4_fwrite(ext4_file *file, const void *buf, size_t size, size_t *wcnt)
 	return r;
 }
 
-int ext4_fseek(ext4_file *file, uint64_t offset, uint32_t origin)
+int ext4_fseek(ext4_file *file, int64_t offset, uint32_t origin)
 {
 	switch (origin) {
 	case SEEK_SET:
-		if (offset > file->fsize)
+		if (offset < 0 || (uint64_t)offset > file->fsize)
 			return EINVAL;
 
 		file->fpos = offset;
 		return EOK;
 	case SEEK_CUR:
-		if ((offset + file->fpos) > file->fsize)
+		if ((offset < 0 && (uint64_t)(-offset) > file->fpos) ||
+		    (offset > 0 &&
```

Comment:
This line is redundant IMO.

Reply:
Example:

offset = -10
file->fpos = 12
file->fsize = 13

This offset is legit. If we omit line 2030:

if ((offset < 0 && (uint64_t)(-offset) > file->fpos)  ||   // evaluates to false
    (uint64_t)offset > (file->fsize - file->fpos))              // evaluates to true
Returns EINVAL

I will add a test case for this.

Reply:
Ups. Sorry, i'm on holiday and I don't think think clearly:)